### PR TITLE
[CDAP-20387] Separate subscriber transaction size from subscriber fetch size

### PR DESCRIPTION
**Pull request by masoud-io**
Why: with the introduction of multiple partitions for ProgramNotificationSubscriberService, and since transaction size is identical to TMS fetch size, multiple concurrent big transactions try to modify DB concurrently which may cause deadlocks. 

This PR, separates txn size from subscriber fetch size so we can still fetch large number of messages from TMS, but can process them in smaller transaction sizes.

### CODE REVIEW

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 1]
```diff
@@ -153,7 +188,32 @@ protected String processMessages(Iterable<ImmutablePair<String, T>> messages) th
       return trackingIterator;
     }, Exception.class);
 
-    return iterator.getLastMessageId();
+    return iterator.lastMessageId;
+  }
+
+  private final class BatchIterator extends AbstractIterator<List> {
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: should have the full generic here, List -> `List<ImmutablePair<String, T>>`

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 2]
```diff
@@ -153,7 +188,32 @@ protected String processMessages(Iterable<ImmutablePair<String, T>> messages) th
...
+    private final MessageTrackingIterator delegate;
+    private final int batchSize;
+
+    private List<ImmutablePair<String, T>> currentBatch;
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: this will result in an extra copy of the data for the default case where fetch size = tx size. 

Instead of doing it this way, we could have a different type of wrapper, something like a LimitIterator, which returns up to n elements. Then create a new LimitIterator for each batch.

Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Added a case to bypass batching if txSize is greater or equal to fetchSize.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 3]
```diff
@@ -134,15 +151,37 @@ protected final String loadMessageId() {
...
     // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
     long timeBoundMillis = 900L * txTimeoutSeconds;
-    iterator = TransactionRunners.run(getTransactionRunner(), context -> {
-      TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
-                                                                                              timeBoundMillis);
-      MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
+    TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
+                                                                                            timeBoundMillis);
+
+    if (this.txSize >= this.getFetchSize()) {
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: this still will result in extra memory usage for some scenarios, it would be better to do something more like:

```
while (timeBoundMessages.hasNext()) {
  LimitIterator iterator = new LimitIterator(timeBoundMessages, this.txSize);
  try {
    lastMessageId = processMessages(context, iterator);
  } ...
}
```

where the LimitIterator is similar to the one we have in NoSqlStructuredTable.

Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: The main challenge is that since txn may be retried, we may have to go over same entries several times. That is why originally, we are using Iterable, and we can call iterator on it inside the txn. This allows us to consume same messages several times inside txn if txn is retried. 
Let me see how I can add this additional logic to LimitIterator

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 4]
```diff
@@ -134,15 +151,37 @@ protected final String loadMessageId() {
...
+      return doProcess(messages);
```
Comment (tivv) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: It seems you are not using time bound here

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 3 above)
Comment (tivv) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: The problem with current approach is that you can't stop in the middle of the batch due to the fact that time is up. As soon as you read the batch, you'll process it to the end even if time is over.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 4 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Fixed

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 3 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 5]
```diff
@@ -280,6 +281,18 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
     getAppMetadataStore(context).persistSubscriberState(getTopicId().getTopic(), "", messageId);
   }
 
+  @Override
+  protected boolean shouldRunInSeparateTx(ImmutablePair<String, Notification> message) {
+    String clusterStatusStr = message.getSecond().getProperties().get(ProgramOptionConstants.CLUSTER_STATUS);
+    if (clusterStatusStr != null &&
+      ProgramRunClusterStatus.valueOf(clusterStatusStr) == ProgramRunClusterStatus.PROVISIONING){
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: space before `{`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 6]
```diff
@@ -280,6 +281,18 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
...
+      // PROVISIONING notifications result in increment in programCountsTable which may result in DB deadlock
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:298]: would be good to also check that the topic is partitioned, as we can't deadlock with just the one partition.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:298]: though if we're doing this, is there a need for the tx size?

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 7]
```diff
@@ -134,23 +159,53 @@ protected final String loadMessageId() {
...
+      if (!messageTrackingIterator.hasNext()) {
+        // message should process in a separate txn
+        messageTrackingIterator = new MessageTrackingIterator(timeBoundIterator);
+      }
+      while (messageTrackingIterator.hasNext() && currentBatch.size() < this.txSize) {
+        currentBatch.add(messageTrackingIterator.next());
+      }
+      try {
+        lastMessageId = doProcess(currentBatch);
+        if (lastMessageId != null && !lastMessageId.equals(messageTrackingIterator.getLastMessageId())) {
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: this is a little hard to understand, maybe would be good to add a stoppedEarly() method to the MessageTrackingIterator?

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 8]
```diff
@@ -46,7 +48,17 @@
...
 
   /**
-   * Constructor.
+   * Specifies tx size for processing fetched messages.
+   * If txSize and fetchSize are identical, then all fetched messages are processed in one transaction.
+   */
+  private final int txSize;
+
+  // Process the notifications and record the message id of where the processing is up to.
+  // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
```
Comment (tivv) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Make it a JavaDoc

Comment (tivv) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Also you don't need to duplicate comments here and in constructor. Here you can simply say it's a time limit and in constructor clarify the calc

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 9]
```diff
@@ -134,23 +159,53 @@ protected final String loadMessageId() {
...
+    List<ImmutablePair<String, T>> currentBatch;
+    String lastMessageId = null;
+    while (timeBoundIterator.hasNext()) {
+      currentBatch = new ArrayList<>();
+      if (!messageTrackingIterator.hasNext()) {
+        // message should process in a separate txn
+        messageTrackingIterator = new MessageTrackingIterator(timeBoundIterator);
+      }
+      while (messageTrackingIterator.hasNext() && currentBatch.size() < this.txSize) {
+        currentBatch.add(messageTrackingIterator.next());
```
Comment (tivv) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: You still has same problem here - you will process the last batch in full even if it goes out of time (assuming a single batch can be processed in time). I would recommend to extract stopwatch and use just a single stopwatch across transactions. Also having two levels of iterators where you wrap twice here and then wrap twice in doProcess makes it hard to follow, e.g. which TrackingIterator is responsible for separating singleton messages into their own transactions?

Comment (tivv) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: I would recommend to reduce number of wrapping iterators, e.g. Tracking Iterator logic can be done on the batching layer. We don't need to refetch, we only need separate transactions

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 6 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:298]: I put this behind a flag. Hope is that realistically when number of pipelines is high, deadlock does not happen. If it happens, then we can reduce txSize to avoid deadlocks. But just in case we start hitting it under very high load with very few pipelines, then we have the option to make this operation into separate txn.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 7 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: stoppedEarly is due to timebound iterator. 
Not needed anymore with the new implementation.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 9 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: We do need tracking iterator because of TimeBound. The confusing part was because of TrackingIterator doing two things: tracking last message and separating transactions if shouldRunInSeparateTx returned true. 
Separated these two logics into different iterators. Should be simpler now.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 10]
```diff
@@ -223,6 +225,11 @@ class ProgramNotificationSingleTopicSubscriberService extends AbstractNotificati
     this.runRecordMonitorService = runRecordMonitorService;
     this.cConf = cConf;
     this.store = store;
+
+    // If number of partitions equals 1, DB deadlock cannot happen as a result of concurrent modifications to
+    // programCountsTable
+    this.checkTxnSeparation = cConf.getInt(Constants.AppFabric.PROGRAM_STATUS_EVENT_NUM_PARTITIONS, 1) > 1
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: don't need a default value in the getter when there is one in cdap-default.xml

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 11]
```diff
@@ -280,6 +287,22 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
     getAppMetadataStore(context).persistSubscriberState(getTopicId().getTopic(), "", messageId);
   }
 
+  @Override
+  protected boolean shouldRunInSeparateTx(ImmutablePair<String, Notification> message) {
+    if (checkTxnSeparation) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: should this be inverted?

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java [hunk 12]
```diff
@@ -351,6 +353,8 @@ public static final class AppFabric {
      */
     public static final String PROGRAM_STATUS_EVENT_NUM_PARTITIONS = "program.status.event.topic.num.partitions";
 
+    public static final String PROGRAM_STATUS_EVENT_CHECK_TXN_SEPARATION = "program.status.event.check.txn.separation";
```
Comment (albertshau) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: nit: let's use tx instead of txn to be consistent with other properties

File: cdap-common/src/main/resources/cdap-default.xml [hunk 13]
```diff
@@ -876,6 +885,17 @@
     </description>
   </property>
 
+  <property>
+    <name>program.status.event.check.txn.separation</name>
```
Comment (albertshau) [inline cdap-common/src/main/resources/cdap-default.xml]: for consistency with other properties, let's add the 'app.' prefix

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 14]
```diff
@@ -134,23 +163,83 @@ protected final String loadMessageId() {
...
+    stopwatch.stop();
+    return lastMessageId;
+  }
+
+  @Nullable
+  private String processSingleTxn(Iterable<ImmutablePair<String, T>> messages,
+                                  Function<Iterable<ImmutablePair<String, T>>,
+                                    MessageTrackingIterator> messageTrackingIteratorFunction)
+    throws Exception {
+    MessageTrackingIterator iterator = TransactionRunners.run(getTransactionRunner(), context -> {
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: would be helpful to add a comment that this method can retry in some situations, which is why messages needs to be an Iterable instead of an Iterator.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 15]
```diff
@@ -134,23 +163,83 @@ protected final String loadMessageId() {
...
+    List<ImmutablePair<String, T>> currentBatch;
+    String lastMessageId = null;
+    while (iterator.hasNext()) {
+      currentBatch = new ArrayList<>();
+      if (!separateTxnIterator.hasNext()) {
+        // next message should process in a separate txn.
+        separateTxnIterator = new SeparateTxnIterator(iterator);
+      }
+
+      while (separateTxnIterator.hasNext() && currentBatch.size() < this.txSize) {
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: with the example input in the javadoc and txSize = 3, doesn't this mean the first batch will contain m1, m2, m3 instead of running m3 in it's own transaction?

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 3 above)
Comment (albertshau) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: This code path stops the iterator early to perform another fetch from the messaging system, while the other code path will run the separate tx in the same while loop. It's not obvious with a quick read of the code, and it's kind of odd to have different behavior in those two scenarios.

I raised the extra memory thing, but that was before understanding that TransactionRunners.run() can be retried. Given the extra complexity that this adds, it's probably better to remove it and handle everything the same way.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 15 above)
Comment (albertshau) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: I feel like we should call shouldRunInSeparateTx() directly in this loop instead of using the SeparateTxnIterator, would be easier to understand the logic

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 16]
```diff
@@ -134,23 +163,83 @@ protected final String loadMessageId() {
...
+      MessageTrackingIterator messageTrackingIterator = messageTrackingIteratorFunction.apply(messages);
+      processMessages(context, messageTrackingIterator);
+      String lastMessageId = messageTrackingIterator.getLastMessageId();
 
       // Persist the message id of the last message being consumed from the iterator
       if (lastMessageId != null) {
         storeMessageId(context, lastMessageId);
       }
-      return trackingIterator;
+      return messageTrackingIterator;
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: nit: a bit more concise to return `messageTrackingIterator.getLastMessageId()`

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 3 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Agree. Most of the complexity was coming from the fact that we wanted to retain both logics. Changed the PR, and it's much simpler.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 15 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: Thanks for your suggestion. Removed SeparateTxnIterator and called shouldRunInSeparateTx directly in the code. It's much more simpler.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java [hunk 17]
```diff
@@ -131,74 +160,114 @@ protected final String loadMessageId() {
...
+      try {
+        String result = processSingleTxn(currentTxMessages, stopwatch);
+        if (result != null) {
+          // we only set lastMessageId if result of current batch is not null
+          // in order to avoid setting lastMessageId to null in case current batch returned null
+          // but previous batch finished.
+          lastMessageId = result;
+        }
+      } catch (Exception ex) {
+        if (lastMessageId == null) {
```
Comment (albertshau) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java:215]: may be good to log if the last message id is not null

**Review by albertshau — APPROVED** [review]
minor comment, otherwise lgtm

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 9 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java]: For completeness on this thread: we ended up removing shouldRunInSeparateTx check from Tracking Iterator, and move it inside the while loop for simplicity.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java (hunk 17 above)
Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java:215]: Done
